### PR TITLE
Generic intermediate nodes fix

### PR DESCRIPTION
**Pull request by mmenozzi**
See https://github.com/symfony-cmf/RoutingAuto/issues/61

### CODE REVIEW

File: Adapter/PhpcrOdmAdapter.php [hunk 1]
```diff
@@ -137,12 +138,24 @@ public function createAutoRoute(UriContext $uriContext, $contentDocument, $autoR
             $parentDocument = $document;
         }
 
+        $finalAutoRoutePath = $path . '/' . $headName;
+        $node = $this->dm->find(null, $finalAutoRoutePath);
+        if ($node) {
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: Blank line before `if` block

File: Adapter/PhpcrOdmAdapter.php [hunk 2]
```diff
@@ -137,12 +138,24 @@ public function createAutoRoute(UriContext $uriContext, $contentDocument, $autoR
...
+            if ($node instanceof Generic) {
+                return $this->convertGenericNodeInAutoRouteNode($node, $contentDocument, $autoRouteTag, $routeType);
+            }
+            $nodeClass = get_class($node);
+            $genericFqcn = 'Doctrine\ODM\PHPCR\Document\Generic';
+            throw new \RuntimeException(
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: Add blank line before `throw`

File: Adapter/PhpcrOdmAdapter.php [hunk 3]
```diff
@@ -137,12 +138,24 @@ public function createAutoRoute(UriContext $uriContext, $contentDocument, $autoR
             $parentDocument = $document;
         }
 
+        $finalAutoRoutePath = $path . '/' . $headName;
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: rename `$path` to `basePath`

File: Adapter/PhpcrOdmAdapter.php [hunk 4]
```diff
@@ -115,6 +115,7 @@ public function removeAutoRoute(AutoRouteInterface $autoRoute)
     public function createAutoRoute(UriContext $uriContext, $contentDocument, $autoRouteTag)
     {
         $path = $this->baseRoutePath;
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: rename to `$basePath`

File: Adapter/PhpcrOdmAdapter.php (hunk 3 above)
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: rename to `$path` if possible after renaming above.

File: Adapter/PhpcrOdmAdapter.php [hunk 5]
```diff
@@ -137,12 +138,24 @@ public function createAutoRoute(UriContext $uriContext, $contentDocument, $autoR
...
+        $finalAutoRoutePath = $path . '/' . $headName;
+        $node = $this->dm->find(null, $finalAutoRoutePath);
+        if ($node) {
+            if ($node instanceof Generic) {
+                return $this->convertGenericNodeInAutoRouteNode($node, $contentDocument, $autoRouteTag, $routeType);
+            }
+            $nodeClass = get_class($node);
+            $genericFqcn = 'Doctrine\ODM\PHPCR\Document\Generic';
+            throw new \RuntimeException(
+                "Unexpected node class '$nodeClass' at path '$finalAutoRoutePath'. Only '$genericFqcn' expected."
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: use `sprintf`, and perhaps reword as follows:

````
sprintf(
    'Encountered existing PHPCR-ODM document at path "%s" of class "%s", the route tree should contain only instances of AutoRouteInterface.'
    $path,
    $documentClass
`)
````

Note that documents that are returned as `Generic` actually are not managed by PHPCR-ODM - the nodes exist in the tree without any ODM mapping.

File: Adapter/PhpcrOdmAdapter.php [hunk 6]
```diff
@@ -137,12 +138,24 @@ public function createAutoRoute(UriContext $uriContext, $contentDocument, $autoR
             $parentDocument = $document;
         }
 
+        $finalAutoRoutePath = $path . '/' . $headName;
+        $node = $this->dm->find(null, $finalAutoRoutePath);
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: rename `$node` to `$document`. We generally use `node` to refer to PHPCR nodes, and `document` to refer to the objects returned from the document manager.

File: Adapter/PhpcrOdmAdapter.php [hunk 7]
```diff
@@ -137,12 +138,24 @@ public function createAutoRoute(UriContext $uriContext, $contentDocument, $autoR
...
+                "Unexpected node class '$nodeClass' at path '$finalAutoRoutePath'. Only '$genericFqcn' expected."
+            );
+        }
         $headRoute = new $this->autoRouteFqcn();
         $headRoute->setContent($contentDocument);
         $headRoute->setName($headName);
         $headRoute->setParent($document);
         $headRoute->setAutoRouteTag($autoRouteTag);
-        $headRoute->setType(AutoRouteInterface::TYPE_PRIMARY);
+        $headRoute->setType($routeType);
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: Revert this .. we only ever set this to TYPE_PRIMARY. Making this dynamic is just confusing?

File: Adapter/PhpcrOdmAdapter.php [hunk 8]
```diff
@@ -190,12 +203,42 @@ public function getReferringAutoRoutes($contentDocument)
     public function findRouteForUri($uri, UriContext $uriContext)
     {
         $path = $this->getPathFromUri($uri);
-
-        return $this->dm->find(null, $path);
+        $node = $this->dm->find(null, $path);
+        if ($node instanceof AutoRouteInterface) {
+            return $node;
+        }
+        return false;
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: Add blank line before return

File: Adapter/PhpcrOdmAdapter.php [hunk 9]
```diff
@@ -190,12 +203,42 @@ public function getReferringAutoRoutes($contentDocument)
     public function findRouteForUri($uri, UriContext $uriContext)
     {
         $path = $this->getPathFromUri($uri);
-
-        return $this->dm->find(null, $path);
+        $node = $this->dm->find(null, $path);
+        if ($node instanceof AutoRouteInterface) {
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: Add blank line before `if` block

File: Adapter/PhpcrOdmAdapter.php [hunk 10]
```diff
@@ -190,12 +203,42 @@ public function getReferringAutoRoutes($contentDocument)
     public function findRouteForUri($uri, UriContext $uriContext)
     {
         $path = $this->getPathFromUri($uri);
-
-        return $this->dm->find(null, $path);
+        $node = $this->dm->find(null, $path);
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: rename `$node` to `$document`

File: Model/AutoRoute.php [hunk 11]
```diff
@@ -50,7 +50,7 @@ public function setType($type)
         $this->setDefault('type', $type);
     }
 
-    public function setRedirectTarget(AutoRouteInterface $redirectRoute)
+    public function setRedirectTarget($redirectRoute)
```
Comment (dantleech) [inline Model/AutoRoute.php]: Is this related?

File: Adapter/PhpcrOdmAdapter.php (hunk 8 above)
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: Return `null` instead of `false` I think? The document manager will return `null` if it found no document.

File: Adapter/PhpcrOdmAdapter.php [hunk 12]
```diff
@@ -190,12 +203,42 @@ public function getReferringAutoRoutes($contentDocument)
...
     }
+
+    /**
+     * @param Generic $node
+     * @param object $contentDocument
+     * @param string $autoRouteTag
+     * @param string $routeType
+     * @return AutoRouteInterface
+     */
+    private function convertGenericNodeInAutoRouteNode(Generic $node, $contentDocument, $autoRouteTag, $routeType)
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: rename to `migrateGenericToAutoRoute` ? Add a comment:

````
Convert the given generic document to an auto route document, migrating any children which it may have.
````

Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: Rename `$node` to `$document`

File: Adapter/PhpcrOdmAdapter.php [hunk 13]
```diff
@@ -190,12 +203,42 @@ public function getReferringAutoRoutes($contentDocument)
...
+    {
+        $autoRouteClassName = $this->autoRouteFqcn;
+        $mapper = $this->dm->getConfiguration()->getDocumentClassMapper();
+        $mapper->writeMetadata($this->dm, $node->getNode(), $autoRouteClassName);
+        $this->dm->getPhpcrSession()->save();
+        // Detach is needed to force Doctrine to re-load the node
+        $this->dm->detach($node);
+        $autoRoute = $this->dm->find(null, $node->getId());
+        if (!$autoRoute instanceof $autoRouteClassName) {
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: Blank line before IF block

File: Adapter/PhpcrOdmAdapter.php [hunk 14]
```diff
@@ -190,12 +203,42 @@ public function getReferringAutoRoutes($contentDocument)
...
+            throw new \RuntimeException(
+                "Something went wrong converting Generic node into an AutoRouteInterface node."
+            );
+        }
+        $autoRoute->setContent($contentDocument);
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: Empty line after IF block

File: Adapter/PhpcrOdmAdapter.php [hunk 15]
```diff
@@ -190,12 +203,42 @@ public function getReferringAutoRoutes($contentDocument)
...
+        $autoRouteClassName = $this->autoRouteFqcn;
+        $mapper = $this->dm->getConfiguration()->getDocumentClassMapper();
+        $mapper->writeMetadata($this->dm, $node->getNode(), $autoRouteClassName);
+        $this->dm->getPhpcrSession()->save();
+        // Detach is needed to force Doctrine to re-load the node
+        $this->dm->detach($node);
+        $autoRoute = $this->dm->find(null, $node->getId());
+        if (!$autoRoute instanceof $autoRouteClassName) {
+            throw new \RuntimeException(
+                "Something went wrong converting Generic node into an AutoRouteInterface node."
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: Use single quotes for exceptions (and all other string assignments)

File: Adapter/PhpcrOdmAdapter.php [hunk 16]
```diff
@@ -190,12 +203,42 @@ public function getReferringAutoRoutes($contentDocument)
...
+            );
+        }
+        $autoRoute->setContent($contentDocument);
+        $autoRoute->setAutoRouteTag($autoRouteTag);
+        $autoRoute->setType($routeType);
+        return $autoRoute;
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: Emoty line before `return`

File: Adapter/PhpcrOdmAdapter.php [hunk 17]
```diff
@@ -190,12 +203,42 @@ public function getReferringAutoRoutes($contentDocument)
...
+     * @param object $contentDocument
+     * @param string $autoRouteTag
+     * @param string $routeType
+     * @return AutoRouteInterface
+     */
+    private function convertGenericNodeInAutoRouteNode(Generic $node, $contentDocument, $autoRouteTag, $routeType)
+    {
+        $autoRouteClassName = $this->autoRouteFqcn;
+        $mapper = $this->dm->getConfiguration()->getDocumentClassMapper();
+        $mapper->writeMetadata($this->dm, $node->getNode(), $autoRouteClassName);
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: @dbu we have discussed migrating document types before -- is this a solution? I seem to remember that it was not so easy.

File: Tests/Functional/EventListener/AutoRouteListenerTest.php [hunk 18]
```diff
@@ -552,4 +552,40 @@ public function testConflictResolverDefaultThrowException()
         $this->getDm()->persist($blog);
         $this->getDm()->flush();
     }
+
+    public function testGenericNodeShouldBeConvertedInAnAutoRouteNode()
+    {
+        $this->assertNull($this->getDm()->find(null, '/test/auto-route/blog'));
+        $this->assertNull($this->getDm()->find(null, '/test/auto-route/blog/my-post'));
```
Comment (dantleech) [inline Tests/Functional/EventListener/AutoRouteListenerTest.php]: We should take for granted that these do not exist. The test should be started in a consistent state.

File: Adapter/PhpcrOdmAdapter.php (hunk 17 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: @dantleech @dbu yes you discussed about that here https://github.com/doctrine/DoctrinePHPCRBundle/issues/92 and here https://github.com/doctrine/phpcr-odm/issues/420. Then I looked at the DocumentMigrateClassCommand (https://github.com/doctrine/phpcr-odm/blob/master/lib/Doctrine/ODM/PHPCR/Tools/Console/Command/DocumentMigrateClassCommand.php) and tried to emulate that. I'm not so expert about Doctrine PHPCR to say if this is a good solution or not but I don't like the alternatives (like rename the original node and create new node).

File: Model/AutoRoute.php (hunk 11 above)
Comment (mmenozzi) [inline Model/AutoRoute.php]: No it's not related. It's only to have a green Travis build.
I don't know if you're aware of it but at this moment if you run tests from master branch of https://github.com/symfony-cmf/RoutingAutoBundle they don't pass because of the change in the AutoRouteInterface (https://github.com/symfony-cmf/RoutingAuto/commit/5d01aa16245b321adfabacfc00abd90021bc0bd8#diff-4f339ec2f4961f0ef07d166e318bf9bf).

File: Adapter/PhpcrOdmAdapter.php (hunk 6 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: A `$document` variable already exists in this scope and is used below. What about `$autoRouteDocument` as new variable name?

Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: how about `$existingDocument` ?

File: Model/AutoRoute.php (hunk 11 above)
Comment (dantleech) [inline Model/AutoRoute.php]: Ah yeah. I have just created a PR to fix that: https://github.com/symfony-cmf/RoutingAutoBundle/pull/166

File: Adapter/PhpcrOdmAdapter.php (hunk 17 above)
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: I think that this is a good solution - so we are basically taking the unmanaged PHPCR node (represented as `Generic` by PHPCR-ODM) and stamping the metadata onto it. @dbu does this make sense to you?

Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: /cc @lsmith77

File: Adapter/PhpcrOdmAdapter.php (hunk 1 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done.

File: Adapter/PhpcrOdmAdapter.php (hunk 2 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done.

File: Adapter/PhpcrOdmAdapter.php (hunk 3 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done.

File: Adapter/PhpcrOdmAdapter.php (hunk 4 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done.

File: Adapter/PhpcrOdmAdapter.php (hunk 5 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done.

File: Adapter/PhpcrOdmAdapter.php (hunk 6 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done.

File: Adapter/PhpcrOdmAdapter.php (hunk 7 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Ok. Done.

File: Adapter/PhpcrOdmAdapter.php (hunk 8 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done.

File: Adapter/PhpcrOdmAdapter.php (hunk 9 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done.

File: Adapter/PhpcrOdmAdapter.php (hunk 10 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done.

File: Adapter/PhpcrOdmAdapter.php (hunk 12 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done.

File: Adapter/PhpcrOdmAdapter.php (hunk 13 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done.

File: Adapter/PhpcrOdmAdapter.php (hunk 14 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done.

File: Adapter/PhpcrOdmAdapter.php (hunk 15 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done.

File: Adapter/PhpcrOdmAdapter.php (hunk 16 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done.

File: Tests/Functional/EventListener/AutoRouteListenerTest.php (hunk 18 above)
Comment (mmenozzi) [inline Tests/Functional/EventListener/AutoRouteListenerTest.php]: Done.

File: Model/AutoRoute.php (hunk 11 above)
Comment (mmenozzi) [inline Model/AutoRoute.php]: Ok. So, you think that I have to revert my commit https://github.com/mmenozzi/RoutingAutoBundle/commit/cf189436c1d8eeacf6e00da4d2757edd642bd46b?

Comment (dantleech) [inline Model/AutoRoute.php]: No, you can just [rebase](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) on master:

````
$ git fetch origin
$ git rebase origin/master
````

File: Adapter/PhpcrOdmAdapter.php [hunk 19]
```diff
@@ -137,6 +137,28 @@ public function createAutoRoute(UriContext $uriContext, $contentDocument, $autoR
...
+            throw new \RuntimeException(
+                sprintf(
+                    'Encountered existing PHPCR-ODM document at path "%s" of class "%s", the route tree should ' .
+                    'contain only instances of AutoRouteInterface.',
+                    $path,
+                    get_class($existingDocument)
+                )
+            );
+        }
         $headRoute = new $this->autoRouteFqcn();
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: Always have a blank line after an if statement

File: Adapter/PhpcrOdmAdapter.php [hunk 20]
```diff
@@ -114,19 +114,19 @@ public function removeAutoRoute(AutoRouteInterface $autoRoute)
      */
     public function createAutoRoute(UriContext $uriContext, $contentDocument, $autoRouteTag)
     {
-        $path = $this->baseRoutePath;
-        $document = $parentDocument = $this->dm->find(null, $path);
+        $basePath = $this->baseRoutePath;
+        $document = $parentDocument = $this->dm->find(null, $basePath);
         if (null === $parentDocument) {
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: Blank line before IF statement

File: Adapter/PhpcrOdmAdapter.php [hunk 21]
```diff
@@ -190,12 +212,49 @@ public function getReferringAutoRoutes($contentDocument)
...
+        return null;
     }
 
     private function getPathFromUri($uri)
     {
         return $this->baseRoutePath . $uri;
     }
+
+    /**
+     * Convert the given generic document to an auto route document, migrating any children which it may have.
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: Now we are not migrating children :) I think we can remove that part of the sentance.

File: Adapter/PhpcrOdmAdapter.php [hunk 22]
```diff
@@ -190,12 +212,49 @@ public function getReferringAutoRoutes($contentDocument)
...
+        $mapper = $this->dm->getConfiguration()->getDocumentClassMapper();
+        $mapper->writeMetadata($this->dm, $document->getNode(), $autoRouteClassName);
+        $this->dm->getPhpcrSession()->save();
+        // Detach is needed to force Doctrine to re-load the node
+        $this->dm->detach($document);
+        $autoRoute = $this->dm->find(null, $document->getId());
+
+        if (!$autoRoute instanceof $autoRouteClassName) {
+            throw new \RuntimeException(
+                'Something went wrong converting Generic node into an AutoRouteInterface node.'
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: We can be more specific here.

````
sprintf('
    Failed to migrate make existing, non-managed, PHPCR node at "%s" to a managed document implementing the AutoRouteInterface. It is an "%s"',
    $document->getId(),
    get_class($autoRoute)
);
````

And a test would be good for this.

**Comment by dantleech** [comment]
Thanks for this, its looking good. Just a few more comments to address and it would be good to add a test for that exception. I would like either @dbu or @lsmith77 to comment on the method used here to migrate the un-managed PHPCR node..

**Comment by WouterJ** [comment]
It also needs a rebase now #166 is merged.

File: Adapter/PhpcrOdmAdapter.php [hunk 23]
```diff
@@ -190,12 +212,49 @@ public function getReferringAutoRoutes($contentDocument)
...
+     *
+     * @param Generic $document
+     * @param object $contentDocument
+     * @param string $autoRouteTag
+     * @param string $routeType
+     * @return AutoRouteInterface
+     */
+    private function migrateGenericToAutoRoute(Generic $document, $contentDocument, $autoRouteTag, $routeType)
+    {
+        $autoRouteClassName = $this->autoRouteFqcn;
```
Comment (lsmith77) [inline Adapter/PhpcrOdmAdapter.php]: one thing that might also work is to detach the document, and then set the ``$autoRouteClassName`` in the find and then save() ...

ie something like:
```
$this->dm->detach($document);
$autoRoute = $this->dm->find($autoRouteClassName, $document->getId());
$this->dm->flush();
```

Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: I didn't understand... You mean that we should remove lines from 241 to 246 and replace them with your three lines?

File: Adapter/PhpcrOdmAdapter.php (hunk 22 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Ok for the exception message.
For the test you mean a unit test to check the exception behavior?

Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: Yeah, just add something like:

````
/**
 * @expectedException \RuntimeException
 * @expectedExceptionMessage some part of the exeption message here
 */
public function testSomeException()
{
}
````

File: Adapter/PhpcrOdmAdapter.php (hunk 23 above)
Comment (lsmith77) [inline Adapter/PhpcrOdmAdapter.php]: yes .. it might work .. basically this would "force" the node to be read into the type of class you want .. and I am hoping that on a flush it will then also update the metadata .. but I am not sure if it will do this .. maybe we should think about explicitly supporting this .. because the below is imho too much replication of PHPCR ODM internals.

Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: I tried and [this test](https://github.com/mmenozzi/RoutingAutoBundle/blob/generic-intermediate-nodes-fix/Tests/Functional/EventListener/AutoRouteListenerTest.php#L556) fails if I do it...

Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: How does it fail? @lsmith77 I would expect that solution to throw an Exception because the Generic document would not be an instance of `$autoRouteClassName` or is `Generic` (unmanaged) a special case? And if so, and that doesn't work, perhaps we should indeed handle it somehow.

Otherwise I am fine with the code here if there is no existing other solution.

Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: @dantleech @lsmith77 I think that the `flush()` call in the `migrateGenericToAutoRoute` causes the `AutoRouteListener` to be called again...

```
1) Symfony\Cmf\Bundle\RoutingAutoBundle\Tests\Functional\EventListener\AutoRouteListenerTest::testGenericNodeShouldBeConvertedInAnAutoRouteNode
Symfony\Cmf\Component\RoutingAuto\ConflictResolver\Exception\ExistingUriException: There already exists an auto route for URL "/blog" and the system is configured to throw this exception in this case. Alternatively you can choose to use a different strategy, for example, auto incrementation. Please refer to the documentation for more information.

/Users/manuele/Work/Projects/RoutingAutoBundle/vendor/symfony-cmf/routing-auto/ConflictResolver/ThrowExceptionConflictResolver.php:32
/Users/manuele/Work/Projects/RoutingAutoBundle/vendor/symfony-cmf/routing-auto/UriGenerator.php:126
/Users/manuele/Work/Projects/RoutingAutoBundle/vendor/symfony-cmf/routing-auto/AutoRouteManager.php:123
/Users/manuele/Work/Projects/RoutingAutoBundle/vendor/symfony-cmf/routing-auto/AutoRouteManager.php:80
/Users/manuele/Work/Projects/RoutingAutoBundle/Doctrine/Phpcr/AutoRouteListener.php:68
/Users/manuele/Work/Projects/RoutingAutoBundle/vendor/symfony/symfony/src/Symfony/Bridge/Doctrine/ContainerAwareEventManager.php:63
/Users/manuele/Work/Projects/RoutingAutoBundle/vendor/doctrine/phpcr-odm/lib/Doctrine/ODM/PHPCR/UnitOfWork.php:3944
/Users/manuele/Work/Projects/RoutingAutoBundle/vendor/doctrine/phpcr-odm/lib/Doctrine/ODM/PHPCR/UnitOfWork.php:2183
/Users/manuele/Work/Projects/RoutingAutoBundle/vendor/doctrine/phpcr-odm/lib/Doctrine/ODM/PHPCR/DocumentManager.php:752
/Users/manuele/Work/Projects/RoutingAutoBundle/Adapter/PhpcrOdmAdapter.php:245
/Users/manuele/Work/Projects/RoutingAutoBundle/Adapter/PhpcrOdmAdapter.php:151
/Users/manuele/Work/Projects/RoutingAutoBundle/vendor/symfony-cmf/routing-auto/AutoRouteManager.php:85
/Users/manuele/Work/Projects/RoutingAutoBundle/Doctrine/Phpcr/AutoRouteListener.php:68
/Users/manuele/Work/Projects/RoutingAutoBundle/vendor/symfony/symfony/src/Symfony/Bridge/Doctrine/ContainerAwareEventManager.php:63
/Users/manuele/Work/Projects/RoutingAutoBundle/vendor/doctrine/phpcr-odm/lib/Doctrine/ODM/PHPCR/UnitOfWork.php:3944
/Users/manuele/Work/Projects/RoutingAutoBundle/vendor/doctrine/phpcr-odm/lib/Doctrine/ODM/PHPCR/UnitOfWork.php:2183
/Users/manuele/Work/Projects/RoutingAutoBundle/vendor/doctrine/phpcr-odm/lib/Doctrine/ODM/PHPCR/DocumentManager.php:752
/Users/manuele/Work/Projects/RoutingAutoBundle/Tests/Functional/EventListener/AutoRouteListenerTest.php:577
```

File: Adapter/PhpcrOdmAdapter.php (hunk 19 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done.

File: Adapter/PhpcrOdmAdapter.php (hunk 20 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done.

File: Adapter/PhpcrOdmAdapter.php (hunk 21 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done.

File: Adapter/PhpcrOdmAdapter.php (hunk 23 above)
Comment (lsmith77) [inline Adapter/PhpcrOdmAdapter.php]: @dantleech afaik we do not write the generic class into the metadata, so from phpcr odm perspective it should not be an issue to change the class .. but if this doesn't work .. maybe we should have a method to migrate the class in PHPCR ODM ..

File: Adapter/PhpcrOdmAdapter.php (hunk 22 above)
Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: Done [here](https://github.com/mmenozzi/RoutingAutoBundle/blob/670ed28373116d684074f5aa6b7354008163f949/Tests/Unit/Adapter/PhpcrOdmAdapterTest.php#L133). What do you think? It's too much tied to the implementation?

File: Adapter/PhpcrOdmAdapter.php (hunk 23 above)
Comment (dbu) [inline Adapter/PhpcrOdmAdapter.php]: > maybe we should have a method to migrate the class in PHPCR ODM ..

we have DocumentMigrateClassCommand - the logic should be refactored
into a helper class that the command simply calls, instead of coding the
logic in the command directly.

Comment (mmenozzi) [inline Adapter/PhpcrOdmAdapter.php]: @dantleech @lsmith77 @dbu I agree. An higher level method which migrates document class should be in Doctrine PHPCR-ODM but I also think that we should merge this PR now in the RoutingAutoBundle and then refactor when this higher level method will be available in Doctrine PHPCR-ODM.

File: Adapter/PhpcrOdmAdapter.php (hunk 22 above)
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: Well, we only need to test up until the Exception -- could we not move the `$autoRoute = $dm->find...` line to the top and only prophesize the document manager?

````
public funcion test...()
{
    $dm->find(null, ...)->willReturn(new \stdClass());
    $this->adapter->createAutoRoute(...);
}
````

(also no need to prophesize `stdClass` - it is part of the PHP lib, and also has no logic)

File: Adapter/PhpcrOdmAdapter.php (hunk 23 above)
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: +1 for keeping this for now

**Comment by dantleech** [comment]
My last comment regarding the exception test was hidden in a diff, repeating here:

Well, we only need to test up until the Exception -- could we not move the $autoRoute = $dm->find... line to the top and only prophesize the document manager?

````
public funcion test...()
{
    $dm->find(null, ...)->willReturn(new \stdClass());
    $this->adapter->createAutoRoute(...);
}
(also no need to prophesize stdClass - it is part of the PHP lib, and also has no logic)
````

File: Adapter/PhpcrOdmAdapter.php (hunk 23 above)
Comment (lsmith77) [inline Adapter/PhpcrOdmAdapter.php]: @dbu I looked at ``DocumentMigrateClassCommand`` and its based on ``NodesUpdateCommand`` .. as such it doesn't make any sense to refactor the command.

Furthermore moving this to a helper class also does not seem to bring all that much benefit since its already all in a dedicated class .. I guess the only benefit would be abstracting the PHPCR session away from the PHPCR ODM user. Since it would still require fetching another class etc not sure if its worth it .. at the same time adding a new method on the DocumentManager for this also seems a bit overboard.

Furthermore, I guess technically what should also work instead of the ``$this->dm->getPhpcrSession()->save();`` call one could also do ``$this->dm->flush($document);`` but even if that works it would have more overhead and could trigger listeners which is likely not intended.

So to conclude, I think the current approach is the best way it can be already. Sorry for the noise :)

Comment (dbu) [inline Adapter/PhpcrOdmAdapter.php]: agreed. and if we want to change on phpcr-odm, we should discuss it
there, not in this PR. sorry from my side as well.

**Comment by mmenozzi** [comment]
@dantleech I'm done with the `stdClass` part of your comment. Now we don't prophesize the `stdClass` and doing sothe `@expectedExceptionMessage` is much simpler.

But I don't get you on the other part of your comment. Remember that with the test `testCreateAutoRouteThrowsExceptionIfItCannotMigrateExistingGenericDocumentToAutoRoute` I want to check the exception in the `migrateGenericToAutoRoute` private method. So I have to prophesize all that stuff which as I said before turns the test much tied with implementation... Let me know.

File: Adapter/PhpcrOdmAdapter.php [hunk 24]
```diff
@@ -190,12 +214,54 @@ public function getReferringAutoRoutes($contentDocument)
     public function findRouteForUri($uri, UriContext $uriContext)
     {
         $path = $this->getPathFromUri($uri);
+        $document = $this->dm->find(null, $path);
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: I think this could equally work with `->dm->find('AutoRouteInterface', $path)` ?

**Comment by dantleech** [comment]
I am not sure myself if what i suggest would work, I will merge this now and make another PR with my idea, thanks for all of this :)

**Comment by dantleech** [comment]
Created: https://github.com/symfony-cmf/RoutingAutoBundle/pull/167

**Comment by mmenozzi** [comment]
You're welcome guys! ;)
Do you know when we can have new stable tag release of RoutingAutoBundle with this implementation? I'm not aware about release process of RoutingAutoBundle. Is anyway related to other Symfony CMF components?

**Comment by WouterJ** [comment]
The Symfony CMF libraries are released mostly at the same time. We're hoping to have a new release in the next 1 or 2 months. This depends a bit on the Doctrine PHPCR-ODM 1.3, which will probably be released next week.

**Comment by dantleech** [comment]
Can we not make a bug fix release for this i.e. 1.0.2?

**Comment by mmenozzi** [comment]
@WouterJ @dantleech I agree for the 1.0.2 release because for my point of view this is the fix for a bug of the 1.0.1 release.

**Comment by WouterJ** [comment]
Then this PR should be cherry-picked/backported into the 1.0 branch as it's now only in master, which will be the new 1.1

**Comment by mmenozzi** [comment]
@WouterJ @dantleech as requested, I've backported the bugfix on the 1.0 branch and opened https://github.com/symfony-cmf/RoutingAutoBundle/pull/168
